### DISCUSSION TITLE
PNG 32 bit floating point

### DISCUSSION THREAD

**Discussion by mrbrklyn** [ImageMagick/ImageMagick#6057] [Development] [answered]
### ImageMagick version

7.1.0-52

### Operating system, version and so on

Linux 

### Description

the gimp is convering png images in 32 bit floating point color, and it reads it fine.  I have a webapp that uses Imagemagik to convert the images into jpgs or other png files and they come out really dark.  I filed a bug report with the gimp.  But they told me the problem will go away if I PNG file to 16 bit int color.  But that still doesn't work and the images on conversion come out very dark

Here is an original for example
 http://www.mrbrklyn.com/coins/1806_half_dollar_0.125.r5_2.o.png

Here is a conversion by imagemagick
http://images.mrbrklyn.com/coins/.cache/1024x1007-1806_half_dollar_0.125.r5_2.o.png

the conversion in the context of the webapp
http://images.mrbrklyn.com/coins/1806_half_dollar_0.125.r5_2.o.png

Here is the same image after png conversion to 16 int in the gimp and then run through the imagemaick web app
http://images.mrbrklyn.com/coins/.cache/1024x1007-1806_half_dollar_0.125.r5_2.o_16int.png
That is not bad but it is sometimes just as dark as the 32 float.

This is the 8bit integer PNG converted to a different sized PNG
http://images.mrbrklyn.com/coins/.cache/1024x1007-1806_half_dollar_0.125.r5_2.o_8int.png

This for example is a PNG 16bit int PNG converted to PNG by imagemagick and it is just too bark
http://images.mrbrklyn.com/coins/west_africa/25_francs_1957_eassai_16bit_int.png

**Comment by fmw42** [comment]
Your PNG file is 16 bit integers per channel for 3 channels and 1 bit alpha which is totally opaque according to IM magick identify -verbose. So essentially your file is 48 bits integers.

Please clarify about it being floats. I do not see that from IM magick identify -verbose, but perhaps IM is reading it in as integers. 

Perhaps save your image to 32-bit floats in TIFF or PFM. I was not aware that PNG supported floats.

When I convert your input in IM to PNG it looks the same.

`magick 1806_half_dollar_0.125.r5_2.o.png 1806_half_dollar_0.125.r5_2.o_im7.png`

```
magick compare -metric rmse 1806_half_dollar_0.125.r5_2.o.png 1806_half_dollar_0.125.r5_2.o_im7.png null:
0 (0)
```

**Reply by mrbrklyn** [reply]
the gimp converts by default all images to 32 bit floating point precision and saves the PNG file as such.  I've been told that PNG files won't hold 32 bit floating point, and that is a bug in the gimp (that they refuse to fix).  So I have to manually convert images to 16 bit integer precision for 8 bit integer precision in order to remain in the PNG specification (although from what I know from coding PNG libraries, I have no idea why PNG cares if it is 32 bit float precision).

**Reply by fmw42** [reply]
Where did you confirm that PNG supports 32-bit floats?

Is that 32-bit floats per channel or 32-bits total for the 3 channels.

I think you confuse processing to floats (IM computes in doubles) with what is saved to the file. I do not know that PNG supports floats. TIFF will and PFM will. Those are the two external formats that IM will support saving to floats.

**Reply by mrbrklyn** [reply]
BTW - a lot of programs on a lot of OSes read a PNG 32 bit float precision without trouble.  Any program that uses ImagMagick libraries for its back end, and there are a lot of them like viewnoir et al, choke on the floating point color.  But why should it have trouble with 16 int bit int precision?

I use PNGs as an archival format as it is lossless but this is making it all but impossible.

**Reply by mrbrklyn** [reply]
http://www.mrbrklyn.com/images/thirtytwo.jpg

32 float precision - that would be on each channel as I understand it.

**Comment by mrbrklyn** [comment]
http://www.mrbrklyn.com/images/thirtytwo_b.jpg

FWIW - same image - to the left in firefox and to the right in Imagemagick - saved in gimp default 32 bit floating point PNG.

**Comment by fmw42** [comment]
That is likely 32-bit float internally, not saved to PNG. IM processes double floats internally, but saves floats only to MIFF, MPC, TIFF and PFM.  The first two are IM formats and the later two are external formats.

Post an image that is 32-bit floats per channel in PNG format. I do not think PNG specs support that.

**Reply by mrbrklyn** [reply]
I appreciate your help, but it is like what I said.  It is being saved with 32 bits against the PNG specification regardless what imagemagick is reporting.  Here is the image that you report as 16 bit as it appears in imagemagick and firefox  

You can see the difference..
http://www.mrbrklyn.com/images/thirtytwo_c.jpg

**Comment by fmw42** [comment]
That was not what I got when converting your input to png. It looked exactly the same as the input, not darker.

What was your IM command?

**Comment by snibgo** [comment]
PNG is integer-only, either 8 or 16 bits (EDIT or 1 or 2 or 4) per channel. Pixel values are 0 to 255, or 0 to 65535, etc. I haven't found any mention of extending PNG to include floating-point values.

The OP links to an image: `http://www.mrbrklyn.com/coins/1806_half_dollar_0.125.r5_2.o.png`. That file has an embedded ICC profile: "GIMP built-in Linear sRGB". Viewers that ignores the profile will mistakenly assume the image is encoded as sRGB, and will display a dark result.

> the gimp converts by default all images to 32 bit floating point precision and saves the PNG file as such.

No. Gimp can not save images as 32-bit floating point PNG.

**Reply by mrbrklyn** [reply]
No. Gimp can not save images as 32-bit floating point PNG. <<==

It seems when I removed all the exif data et all, it does work correctly.

**Reply by mrbrklyn** [reply]
<<<The OP links to an image: http://www.mrbrklyn.com/coins/1806_half_dollar_0.125.r5_2.o.png. That file has an embedded ICC profile: "GIMP built-in Linear sRGB". Viewers that ignores the profile will mistakenly assume the image is encoded as sRGB, and will display a dark result.>>

Can you explain that in more detail

**Reply by mrbrklyn** [reply]
I am very excited that this seems to work now that I strip the extra data when saving the file.  I feel like I've been suffering with this for a few years now.  My guess is that putting the color profile into the file caused viewers including imagemagick to fail to read the PNG file correctly.  but as I understand it ICCs are important.

**Comment by snibgo** [comment, marked as answer]
On PNG sample precision, see https://en.wikipedia.org/wiki/PNG and http://www.libpng.org/pub/png/spec/1.2/ .

PNG files do not record samples as floating-point. Each sample is an integer of 1, 2, 4, 8 or 16 bits.

Gimp can store an image in memory with samples as 32-bit floats. But when the image is exported to a file, Gimp will convert the image to whatever the file format can record.

An ICC color profile defines the colorimetric meanings of the pixel samples. A common profile, "sRGB", has lightness values adjusted to roughly match human vision, and also roughly to roughly match old-fashioned cathode-ray tube displays. By contrast, "linear RGB" has values that are (or should be) proportional to the energy of light the values represent.

sRGB and linear RGB use the same primaries (ie the same meaning of "red", "green" and "blue"), but only sRGB has a transfer response curve.

See, for example, https://en.wikipedia.org/wiki/RGB_color_spaces

ImageMagick should have no problem reading files with embedded ICC colour profiles, and processing those images. See the "-profile" operation.

**Reply by mrbrklyn** [reply]
Thanks

There is something being miscommunicated between the gimp and imagemagick when saving as default in the gimp.  Maybe the gimp is saying it is putting in an ICC and then failing to do so?  If I give a broken file do we have a means to examine the PNG chuncks and see where the error is.  If it is with the gimp people, I can point and say - HERE is the trouble.  They are currently annoyed with me for bothering them about this adnuseum.   I wrote a C++ program that read PNG files about 8 years back.  http://www.nylxs.com/docs/png_project_presentation.html

**Comment by fmw42** [comment]
What processing are you doing in GIMP?  You seem to be attaching a linear RGB profile. Perhaps that is a default and you do not know it is happening. Mostly we use sRGB profiles for output, but you may be doing some processing purposely in a linear RGB space. So GIMP assigns a linear profile.

**Comment by fmw42** [comment]
Your issue may just be the viewer that you are using to view the results. Not all viewer will use the profile that is imbedded in your image meta data. They often assume sRGB. So if you are using a linear view and they do not use it, it could look wrong to you.

**Reply by mrbrklyn** [reply]
Your issue may just be the viewer <<===

The view is imagemagick

**Comment by fmw42** [comment]
If you strip the profile, the result will be dark. Try

`magick 1806_half_dollar_0.125.r5_2.o.png +profile "*" dark.png`

If you convert the profile from linear to sRGB, it looks the same as I see your input with a viewer that respects your profile.

 `magick 1806_half_dollar_0.125.r5_2.o.png -profile /Users/fred/images/profiles/sRGB.icc srgb.png`

So I think that explains your issue. You are using a viewing that is not respecting or stripping your profile.

**Reply by mrbrklyn** [reply]
If you strip the profile, the result will be dark. >>>

FWIW - I stripped the profile and it lighten up
This is the fixed image now
http://www.mrbrklyn.com/coins/1806_half_dollar_0.125.r5_2.o.png

This is it now fixed through the imagemagick filter
http://images.mrbrklyn.com/coins/1806_half_dollar_0.125.r5_2.o.png

This last one, when I has the profile in it, was very dark.

**Comment by fmw42** [comment]
Your image is dark when display in IM because it respect the linear profile. When you change the profile to sRGB it should look light again (i.e. normal). When you strip the profile, PNG signs an sRGB profile to it.  JPG may behave differently.

You would be less confused it you always convert to sRGB profile.

**Reply by mrbrklyn** [reply]
IM is reading the profile but the gimp and darktable and firefox is not?  The problem is deeper than that.  It is like I said.  Instead of speculation - we need to look at a file that behaves like this and actually see what it says and then we can KNOW why it is misbehaving.
